returnn/util/sig_proc.py

```diff
@@ -0,0 +1,159 @@
+  https://en.wikipedia.org/wiki/Greenwood_function
+  The default values are taken from RASR, see https://github.com/rwth-i6/rasr/blob/master/src/Signal/GammaTone.cc
+  They correspond to the recommended values for human data according to
+  Greenwood, Donald D. "A cochlear frequency‐position function for several species—29 years later."
+  The Journal of the Acoustical Society of America, 1990
+
+  :param float freq: frequency
+  :param float scaling_constant: A in [1]
+  :param float constant_of_integration: k in [1]
+  :param float slope: a in [1]
+  :return: fractional length corresponding to given frequency
+  :rtype: float
+  """
+  return numpy.log10(freq / scaling_constant + constant_of_integration) / slope
+
+
+class GammatoneFilterbank(object):
+  """
+  Class representing a gammatone filterbank.
+  Based on
+  [1] Schlueter, Ralf, et al. "Gammatone features and feature combination for large vocabulary speech recognition."
+  ICASSP 2007
+  """
+
+  def __init__(self, num_channels, length, sample_rate=16000, freq_max=7500., freq_min=100., normalization=True):
+    """
+    :param int num_channels: number of filters
+    :param int length: length of FIR filters in seconds
+    :param int sample_rate: sample rate of audio signal in Hz
+    :param float freq_max: maximum frequency of filterbank
+    :param float freq_min: minimum frequency of filterbank
+    :param bool normalization: normalize filterbanks to maximum frequency response of 0 dB
+    """
+    self.num_channels = num_channels
+    self.length = length
+    self.sample_rate = sample_rate
+    self.freq_max = freq_max
+    self.freq_min = freq_min
+    self.normalization = normalization
+
+  def get_gammatone_filterbank(self):
+    """
+    Returns an array with the parameters of the gammatone filterbank
+
+    :return: gammatone filterbank
+    :rtype: numpy.array
+    """
+    center_freqs = self.center_frequencies(self.num_channels, self.freq_max, self.freq_min)
+    fbank = []
+    for freq in center_freqs:
+      fbank.append(self.gammatone_impulse_response(freq, self.length, self.sample_rate))
+    fbank = numpy.vstack(fbank)
+    if self.normalization:
+      fbank = self.normalize_filters(fbank)
+    fbank = numpy.transpose(fbank)
+    return fbank
+
+  @staticmethod
+  def center_frequencies(num_channels, freq_max, freq_min):
+    """
+    Determine center frequencies for gammatone filterbank
+
+    :param int num_channels: number of filters
+    :param float freq_max: maximum frequency of filterbank
+    :param float freq_min: minimum frequency of filterbank
+    :return: center frequencies
+    :rtype: numpy.array
+    """
+    x_min = inv_greenwood_function(freq_min)
+    x_max = inv_greenwood_function(freq_max)
+    x_linear = numpy.linspace(x_min, x_max, num_channels)
+    freq_center = numpy.vectorize(greenwood_function)(x_linear)
+    return freq_center
+
+  @staticmethod
+  def bandwidth_by_center_frequency(freq, l_=24.7, q_=9.264491981582191):
```

Comment:
Why `l_` and `q_` (with `_` postfix)?

Reply:
I took the names from RASR and kept the postfix because PyCharm issues a warning if a variable is just named `l`. I could rename `q_` into `quality_factor`, but for `l_` I don't know a good name. I checked the paper and couldn't find a descriptive name in there.

Reply:
Yes, because PyCharm is right, you should not use `l` and `q`. And using `_` is not really a solution. Name it `a` and `b` or `x` and `y` if you really think there are no good names for these. Or `x` and `quality_factor`. But maybe you have a better idea what this `x` is about.